com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/EvaluateRequestHandler.java

```diff
@@ -54,130 +54,56 @@
+                long threadId = stackFrameReference.getThread().uniqueID();
+                if (value instanceof ObjectReference) {
+                    VariableProxy varProxy = new VariableProxy(stackFrameReference.getThread(), "eval", value);
+                    int referenceId = VariableUtils.hasChildren(value, showStaticVariables)
+                            ? context.getRecyclableIdPool().addObject(threadId, varProxy) : 0;
+                    int indexedVariableId = value instanceof ArrayReference ? ((ArrayReference) value).length() : 0;
+                    response.body = new Responses.EvaluateResponseBody(variableFormatter.valueToString(value, options),
+                            referenceId, variableFormatter.typeToString(value == null ? null : value.type(), options),
+                            indexedVariableId);
+                    return response;
+                }
+                // for primitive value
+                response.body = new Responses.EvaluateResponseBody(variableFormatter.valueToString(value, options), 0,
+                        variableFormatter.typeToString(value == null ? null : value.type(), options), 0);
+                return response;
+            } catch (InterruptedException | ExecutionException e) {
+                Throwable cause = e;
+                if (e instanceof ExecutionException && e.getCause() != null) {
+                    cause = e.getCause();
+                }
+                // TODO: distinguish user error of wrong expression(eg: compilation error)
+                logger.log(Level.WARNING, String.format("Cannot evalution expression because of %s.", cause.toString()), cause);
+                throw new CompletionException(cause);
             }
-            currentValue = obj.getValue(field);
-        }
-
-        int referenceId = 0;
-        if (currentValue instanceof ObjectReference && VariableUtils.hasChildren(currentValue, showStaticVariables)) {
-            // save the evaluated value in object pool, because like java.lang.String, the evaluated object will have sub structures
-            // we need to set up the id map.
-            VariableProxy varProxy = new VariableProxy(thread.uniqueID(), "Local", currentValue);
-            referenceId = context.getRecyclableIdPool().addObject(thread.uniqueID(), varProxy);
-        }
-        int indexedVariables = 0;
-        if (currentValue instanceof ArrayReference) {
-            indexedVariables = ((ArrayReference) currentValue).length();
-        }
-        response.body = new Responses.EvaluateResponseBody(context.getVariableFormatter().valueToString(currentValue, options),
-                referenceId, context.getVariableFormatter().typeToString(currentValue == null ? null : currentValue.type(), options),
-                indexedVariables);
-        return CompletableFuture.completedFuture(response);
+        });
+        return completableFuture;
```

Comment:
Simply return when the future is completed. No need to allocate a variable and do late return.